docs/rules/no-implicit-injections.md

```diff
@@ -0,0 +1,112 @@
+# ember/no-implicit-injections
+
+🔧 This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).
+
+<!-- end auto-generated rule header -->
+
+Ember 3.26 introduced a deprecation for relying on implicit service injections or allowing addons to implicitly inject services into all classes of certain types. Support for this is dropped in Ember 4.0.
+
+In many applications, `this.store` from Ember Data is often used without injecting the `store` service in Controllers or Routes. Other addons may also have included implicit service injections via initializers and the `application.inject` API.
+
+To resolve this deprecation, a service should be explicitly declared and injected using the [service injection decorator](https://api.emberjs.com/ember/3.28/functions/@ember%2Fservice/inject).
+
+## Rule Details
+
+This rule checks for a configured list of previously auto injected services and warns if they are used in classes without explicit injected service properties.
+
+## Examples
+
+Examples of **incorrect** code for this rule:
+
+```js
+// routes/index.js
+import Route from '@ember/routing/route';
+
+export default class IndexRoute extends Route {
+  model() {
+    return this.store.findAll('rental');
+  }
+}
+
+```
+
+```js
+// controllers/index.js
+import Controller from '@ember/controller';
+import { action } from '@ember/object';
+
+export default class IndexController extends Controller {
+  @action
+  loadUsers() {
+    return this.store.findAll('user');
+  }
+}
+```
+
+Examples of **correct** code for this rule:
+
+```js
+// routes/index.js
+import Route from '@ember/routing/route';
+import { inject as service } from '@ember/service';
+
+export default class IndexRoute extends Route {
+  @service('store') store;
+
+  model() {
+    return this.store.findAll('rental');
+  }
+}
+```
+
+```js
+// controller/index.js
+import Route from '@ember/routing/route';
+import { action } from '@ember/object';
+import { inject as service } from '@ember/service';
+
+export default class IndexController extends Controller {
+  @service('store') store;
+
+  @action
+  loadUsers() {
+    return this.store.findAll('user');
+  }
+}
+```
+
+## Migration
+
+The autofixer for this rule will update classes and add injections for the configured services.
+
+## Configuration
+
+This lint rule will search for instances of `store` used in routes or controllers by default. If you have other services that you would like to check for uses of, the configuration can be overridden.
+
+- object -- containing the following properties:
+  - array -- `services` -- Array of configuration objects configuring the lint rule to check for use of implicit injected services
+    - string -- `serviceName` -- The name of the service that is implicitly injected
```

Comment:
https://github.com/ember-cli/eslint-plugin-ember/pull/1714/files/#diff-ce4c727cc50236f1ac7ea5af2f03053469b2bff730a010ed6f0c25d29ab97319R90-R93

I added the check for rule behavior as requested and updated the md to reference that service names should be kebab cased. https://github.com/ember-cli/eslint-plugin-ember/pull/1714/files/#diff-ce4c727cc50236f1ac7ea5af2f03053469b2bff730a010ed6f0c25d29ab97319R90-R93

I'm not sure I follow what you're saying in the bullet list and I'm unclear on `isInjectedServiceProp` comment. I based things off the work in `no-assigment-of-untracked` and the patterns there since I found it to most closely resemble the state checks required to look for uses.

Reply:
The part about the input format for service names has been addressed.

But I still want to make sure every format of service name mentioned above is able to be detected by your rule when it checks if a service injection already exists. Can you add a test case checking for all possible formats of service injections (including nested service names) to ensure the rule is able to detect each one? You could just pass in ~10 service names to the denylist for one test case and then inject each one in a different format to ensure it's detected.

I was just suggesting `isInjectedServiceProp` since it could be helpful in regards to checking for service injections and handling the different formats, but not required if you already have a good way to do it.

Reply:
The service definition types aren't allowed in the deny list. The deny list config has to be the service module path.

I did however add tests for ways service could be declared in a class and added/fixed use in legacy `.extends` modules

Reply:
Got it. Thanks.

Have you tested nested services yet? The service name could look like `my-scope/my-service`.

Reply:
Though...

Now I'm realizing that `service` alone isn't really 100% coverage for uses of auto-injected services. 

The current config and code assumes a 1:1 mapping (👋🏽 around kebab and camelCasing) of the injected property name to the service name. But... This may not be the case: initializers can inject any property name they want:

```js
// Example of 1:1 mapping
application.inject('controller', 'flashMessages', 'service:flash-messages');

// Example of arbitrary mapping
application.inject('controller', 'somethingElse', 'service:flash-messages');
```

I think the config feels good for the 80% usecase (and TBH most teams will likely not configure at all and just look for implicit use of `store`.

We could augment this with a config `propertyName` which would default to `camelCase(config.service)`

Reply:
If we care about that edge case at all, we can worry about it as a follow-up. I'm ready to merge this PR as soon as I am sure nested services are handled.

Reply:
The thought of `propertyName` came from the issue with nested services`. Ember AFAIK doesn't have a way of auto mapping a property to a nested service.

So if you have an autoInjection of `@service('cart/checkout')` there has to be an explicit `propertyName`. I'll add an assertion to check for this config.

Reply:
Okay. I would still like to see a test with the nested service name.

Reply:
Updated and added the assertion mentioned above